README.md
bolt
====

Userspace system daemon to enable security levels for *Thunderbolt™*
on GNU/Linux®.

Introduction
------------

Thunderbolt™ is the brand name of a hardware interface developed by
Intel® that allows the connection of external peripherals to a
computer.

Devices connected via Thunderbolt can be DMA masters and thus read
system memory without interference of the operating system (or even
the CPU). Version 3 of the interface introduced 5 different security
levels, in order to mitigate the aforementioned security risk that
connected devices pose to the system. The security level is set by the
system firmware.

The five security levels are:

 * `none`:    Security disabled, all devices will fully functional
              on connect.
 * `dponly`:  Only pass the display-port stream through to the
              connected device.
 * `user`:    Connected devices need to be manually authorized by
              the user.
 * `secure`:  As 'user', but also challenge the device with a secret
              key to verify its identity.
 * `usbonly`: One PCIe tunnel is created to a usb controller in a
              thunderbolt dock; no other downstream PCIe tunnels are
              authorized (needs 4.17 kernel and recent hardware).

The Linux kernel, starting with version 4.13, provides an interface via
sysfs that enables userspace query the security level, the status of
connected devices and, most importantly, to authorize devices, if the
security level demands it.

boltd - the system daemon
-------------------------

The core of bolt is a system daemon (`boltd`) that interfaces with
sysfs and exposes devices via D-Bus to clients. It also has a database
of previously authorized devices (and their keys) and will, depending
on the policy set for the individual devices, automatically authorize
newly connected devices without user interaction. The daemon supports
syncing the devices database with the pre-boot access control list
firmware feature. It also adapts its behavior when iommu support is
detected.

boltctl - command line client
-----------------------------
The `boltctl` command line can be used to manage thunderbolt devices
via `boltd`.  It can list devices, monitor changes and initiate
authorization of devices.


Installation
============

The [meson][meson] build system is used to configure and compile bolt.


    meson build           # configure bolt, use build as buildir
    ninja -C build        # compile it
    ninja -C build test   # run the tests

See [INSTALL][install] for more information, [BUGS][bugs] for how to
file issues and [HACKING][hacking] how to contribute.


[meson]: http://mesonbuild.com/
[install]: INSTALL.md
[bugs]: BUGS.md
[hacking]: HACKING.md


CHANGELOG.md
Version 0.9.3
-------------
_Fix for the fix_
Released: 2022-08-23

* Work around a glib fix that changed behavior of the log API where now
  calling `g_log_set_writer_func` results in an hard error, which broke
  the test suite since aforementioned function was indeed called more
  than once. Nothing a level of indirection couldn't fix.

* CI: disable Debian based ci for now. Another glib fix broke the Debian
  and Arch based docker container. Fedora has patched glib to bake the
  fix out again so it is fine.


Version 0.9.2
-------------
_Please get along_
Released: 2022-01-27

* This release is compatible with umockdev >= 0.16.3; there was a change in
  umockdev that made our test fail with it, since both our test and umockev
  were trying to create the same directorires. bolt now allows for the dir
  to already exist.

* The license for `90-bolt.rules` has changed from `GPL-2.1+`, which does
  not exist and was probably was confused with `LGPL-2.1+`, to `GPL 2.0+`.

* Documentation has been updated and spelling mistakes fixed.

* Various improvements for continuous integration.

* The minimum required version of meson has been bumped to 0.46.0.


Version 0.9.1
-------------
_Unstable icy waters_
Released: 2020-11-30

* Bug fixes for integrated thunderbolt controllers:
  On Ice Lake, the Thunderbolt 3 i/o subsystem is fully integrated into the die.
  As a side effect it does not have a DROM, which means the host udev device
  does not have the device and vendor name and id attributes.
  Additionally the `unique_id` of said host controller changes with every boot,
  which breaks one of the fundamental assumptions in `boltd`. Therefore a number
  of bug fixes were necessary to properly support this new architecture:

  - Don't store domains where uuids change across reboots [!220]
  - Fixes for the journal and the domain's acl-log [!221]
  - Version the store and use that to clean up stale domains once [!226, !231]
  - Host identification for embedded thunderbolt controllers [!233]

* Various other small bug fixes and memory leak fixes.


Version 0.9
-----------
_Four comes after Three_
Released: 2020-06-15

* New Features:
  - Add 'Generation' attribute for the Manager [!197]
  - Ability to change the policy of a stored device [!202]
  - The BootACL Domain property is now writable [!184]
  - Support for systemd's service watchdog [!185]
  - Expose Link Speed sysfs attributes [!214]

* Improvements:
  - boltclt: show timestamps in 'monitor' call [!208]
  - Persist the host device [!194]

* Bug fixes:
  - Fix a flaky test [!217, #161]
  - Plug small memory leaks in error conditions [!217]
  - Ignore spurious wakeup device uevents for probing [!209]
  - Preserve keystate when updating devices [!192]


Version 0.8
-----------
_I owe it to the MM U!_
Released: 2019-06-14

* New Features:
  - **IOMMU support**: adapt behavior iommu support is present and active [#128]
    - automatically enroll new devices with the new `iommu` policy when iommu is active
    - automatically authorize devices with the `iommu` policy if iommu is active
  - `boltctl config` command to describe, get and set global, device and domain properties.
  - Chain authorization and enrollment via `boltctl {enroll, authorize} --chain` [!153, !154]
  - `bolt-mock` script for interactively testing `boltd` [!152]

* Improvements:
  - Automatically import devices that were authorized at boot [#137]
  - Make tests installable [#140]
  - Honour `STATE_DIRECTORY` [!159] and `RUNTIME_DIRECTORY` [!161]
  - Profiling support via gprof [!168]

* Bug fixes:
  - Better handling of random data generation [#132, !165]
  - Fix double free in case of client creation failure [!148]
  - Fix invalid format string in warning [!14]

* NB for packagers:
  - The dbus configuration is now installed in `$datadir/dbus-1/system.d` instead of `$sysconfdir` [!177].
  - To install tests, configure with `-Dinstall-tests=true`.


Version 0.7
-----------
_The Known Unknowns_
Released: 2019-01-01


* Features:
  - announce status to systemd via sd_notify (using a simple custom implementation) [!143]

* Bug fixes:
  - properly update global security level status [#131 via !141]
  - adapt to `systemd` 240 not sending `bind`/`unbind` uevents [#133 via !145]
  - fix compilation on musl [#126 via !140]
  - daemon: use `g_unix_signal_source…` to catch signals [#127, #129 via !138]

* Improvements
  - precondition checks cleanup and completion [#124 via !139]
  - error cleanup [#125, !142]
  - fix some leaks and issues uncovered by coverity [!144]


Version 0.6
-----------
_Make the firmware do it!_
Released: 2018-11-28

* New Features:
  - **pre-boot access control list, aka. `BootACL`** support [!119]
    - domains objects are now persistent
      - new `Uid` (dbus) / `uid` (object) property derived from the uuid of the device representing the root switch
      - `sysfs` and `id` attribute will be set/unset on connects and disconnects
      - domains are now stored in the boltd database
    - domains got the `BootACL` (dbus) / `bootacl` (object) property
      - uuids can be added, removed or set in batch
      - when domain is *online*: changes are written to the sysfs `boot_acl` attribute directly
      - when domain is *offline*: changes are written to a journal and then reapplied in order when the domain is connected
    - newly enrolled devices get added to all bootacls of all domains *if* the `policy` is `BOLT_POLICY_AUTO`
    - removed devices get deleted from all bootacls of all domains
    - `boltacl domain` command will show the bootacl slots and their content

  - `boltctl` gained the `-U, --uuid` option, to control how uuids are printed [!124]

* Improvements and fixes:
  - Testing [!127]
    - The test coverage increased to `84.80%` overall and to `90.0%` for the `boltd` source
    - Coverage is reported for merge requests via the fedora ci image [!126]
    - `boltctl` is now included in the tests [!132]

cli/boltctl.c
/*
 * Copyright © 2017 Red Hat, Inc
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library. If not, see <http://www.gnu.org/licenses/>.
 *
 * Authors:
 *       Christian J. Kellner <christian@kellner.me>
 */

#include "config.h"

#include "boltctl-cmds.h"
#include "boltctl-uidfmt.h"

#include "bolt-client.h"
#include "bolt-enums.h"
#include "bolt-error.h"
#include "bolt-macros.h"
#include "bolt-str.h"
#include "bolt-term.h"
#include "bolt-time.h"

#include <gio/gio.h>

#include <locale.h>
#include <stdlib.h>

gboolean
check_argc (int      argc,
            int      lower,
            int      upper,
            GError **error)
{
  argc--; /* discard argv[0] */

  if (lower == upper && argc != upper)
    {
      g_set_error (error, BOLT_ERROR, BOLT_ERROR_FAILED,
                   "unexpected number of arguments: %d, wanted %d",
                   argc, upper);
      return FALSE;
    }
  else if (argc < lower)
    {
      g_set_error (error, BOLT_ERROR, BOLT_ERROR_FAILED,
                   "not enough arguments: %d, wanted at least %d",
                   argc, lower);
      return FALSE;
    }
  else if (argc > upper)
    {
      g_set_error (error, BOLT_ERROR, BOLT_ERROR_FAILED,
                   "too many arguments: %d, wanted at most %d",
                   argc, lower);
      return FALSE;
    }

  return TRUE;
}

int
usage_error (GError *error)
{
  g_printerr ("%s:", g_get_application_name ());
  g_printerr ("%s error: %s", bolt_color (ANSI_RED), bolt_color (ANSI_NORMAL));
  if (error)
    g_printerr ("%s", error->message);
  g_printerr ("\n");
  g_printerr ("Try \"%s --help\" for more information.", g_get_prgname ());
  g_printerr ("\n");

  return EXIT_FAILURE;
}

int
usage_error_need_arg (const char *arg)
{
  g_autoptr(GError) error = NULL;

  g_set_error (&error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
               "missing argument '%s'", arg);
  return usage_error (error);
}

int
usage_error_too_many_args (void)
{
  g_autoptr(GError) error = NULL;

  g_set_error (&error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
               "too many arguments");
  return usage_error (error);
}

int
report_error (const char *prefix, GError *error)
{
  g_printerr ("%s:", g_get_application_name ());
  g_printerr ("%s error: %s", bolt_color (ANSI_RED), bolt_color (ANSI_NORMAL));

  if (prefix)
    g_printerr ("%s", prefix);

  if (error)
    {
      if (g_dbus_error_is_remote_error (error))
        g_dbus_error_strip_remote_error (error);

      if (prefix)
        g_printerr (": ");

      g_printerr ("%s", error->message);
    }

  g_printerr ("\n");

  return EXIT_FAILURE;
}


/* device related commands */
static gboolean
format_timestamp (BoltDevice *dev,
                  char       *buffer,
                  gsize       n,
                  const char *timesel)
{
  g_autofree char *str = NULL;
  guint64 ts;

  g_object_get (dev, timesel, &ts, NULL);

  if (ts > 0)
    {
      str = bolt_epoch_format (ts, "%c");
      g_utf8_strncpy (buffer, str, n);
    }
  else
    {
      g_utf8_strncpy (buffer, "no", n);
    }

  return ts > 0;
}

static gboolean
format_generation (guint generation, char *buffer, size_t n)
{
  switch (generation)
    {
    case 1:
    case 2:
    case 3:
      g_snprintf (buffer, n, "Thunderbolt %u", generation);
      return TRUE;

    case 4:
      g_snprintf (buffer, n, "USB4");
      return TRUE;

    case 0:
      g_snprintf (buffer, n, "Unknown");
      return FALSE;

    default:
      g_snprintf (buffer, n, "%u", generation);
      return TRUE;
    }
}

void
print_device (BoltDevice *dev, gboolean verbose)
{
  g_autofree char *label = NULL;
  const char *path;
  const char *uid;
  const char *name;
  const char *vendor;
  const char *syspath;
  const char *parent;
  BoltDeviceType type;
  BoltStatus status;
  BoltAuthFlags aflags;
  BoltKeyState keystate;
  BoltPolicy policy;
  const char *status_color;
  const char *status_symbol;
  const char *type_text;
  const char *status_text;
  const char *tree_branch;
  const char *tree_right;
  const char *tree_space;
  gboolean stored;
  gboolean pcie;
  char buf[256];
  guint gen;

  path = g_dbus_proxy_get_object_path (G_DBUS_PROXY (dev));
  uid = bolt_device_get_uid (dev);
  name = bolt_device_get_name (dev);
  vendor = bolt_device_get_vendor (dev);
  gen = bolt_device_get_generation (dev);
  type = bolt_device_get_device_type (dev);
  status = bolt_device_get_status (dev);
  aflags = bolt_device_get_authflags (dev);
  parent = bolt_device_get_parent (dev);
  syspath = bolt_device_get_syspath (dev);
  stored = bolt_device_is_stored (dev);
  policy = bolt_device_get_policy (dev);
  keystate = bolt_device_get_keystate (dev);

  pcie = bolt_flag_isclear (aflags, BOLT_AUTH_NOPCIE);

  status_symbol = bolt_glyph (BLACK_CIRCLE);
  tree_branch = bolt_glyph (TREE_BRANCH);
  tree_right = bolt_glyph (TREE_RIGHT);
  tree_space = bolt_glyph (TREE_SPACE);

  switch (status)
    {
    case BOLT_STATUS_DISCONNECTED:
      status_symbol = bolt_glyph (WHITE_CIRCLE);
      status_color = bolt_color (ANSI_NORMAL);
      status_text = "disconnected";
      break;

    case BOLT_STATUS_CONNECTING:
      status_color = bolt_color (ANSI_YELLOW);
      status_text = "connecting";
      break;

    case BOLT_STATUS_CONNECTED:
      status_color = bolt_color (ANSI_YELLOW);
      status_text = "connected";
      break;

    case BOLT_STATUS_AUTHORIZED:
    case BOLT_STATUS_AUTHORIZED_NEWKEY:
    case BOLT_STATUS_AUTHORIZED_SECURE:
    case BOLT_STATUS_AUTHORIZED_DPONLY:
      if (pcie)
        {
          status_color = bolt_color (ANSI_GREEN);
          status_text = "authorized";
        }
      else
        {
          status_color = bolt_color (ANSI_BLUE);
          status_text = "connected (no PCIe tunnels)";
        }
      break;

    case BOLT_STATUS_AUTH_ERROR:
      status_color = bolt_color (ANSI_RED);
      status_text = "authorization error";
      break;

    default:
      status_color = bolt_color (ANSI_NORMAL);
      status_text = "unknown";
      break;
    }

  label = bolt_device_get_display_name (dev);

  g_print (" %s%s%s %s\n",
           status_color,
           status_symbol,
           bolt_color (ANSI_NORMAL),
           label ? : name);

  type_text = bolt_device_type_to_string (type);

  g_print ("   %s type:          %s\n", tree_branch, type_text);
  g_print ("   %s name:          %s\n", tree_branch, name);
  g_print ("   %s vendor:        %s\n", tree_branch, vendor);
  g_print ("   %s uuid:          %s\n", tree_branch, format_uid (uid));
  if (verbose)
    g_print ("   %s dbus path:     %s\n", tree_branch, path);
  if (format_generation (gen, buf, sizeof (buf)) || verbose)
    g_print ("   %s generation:    %s\n", tree_branch, buf);
  g_print ("   %s status:        %s\n", tree_branch, status_text);

  if (bolt_status_is_connected (status))
    {
      g_autofree char *flags = NULL;
      const char *domain;
      BoltLinkSpeed speed = {0, };

      domain = bolt_device_get_domain (dev);
      g_print ("   %s %s domain:     %s\n",
               bolt_glyph (TREE_VERTICAL),
               tree_branch,
               domain);

      if (verbose)
        {
          g_print ("   %s %s parent:     %s\n",
                   bolt_glyph (TREE_VERTICAL),
                   tree_branch,
                   parent);
          g_print ("   %s %s syspath:    %s\n",
                   bolt_glyph (TREE_VERTICAL),
                   tree_branch,
                   syspath);
        }

      bolt_device_get_linkspeed (dev, &speed);
      if (speed.rx.lanes && speed.rx.speed)
        {
          g_print ("   %s %s rx speed:   %u Gb/s = %u lanes * %u Gb/s\n",
                   bolt_glyph (TREE_VERTICAL),
                   tree_branch,
                   speed.rx.lanes * speed.rx.speed,
                   speed.rx.lanes, speed.rx.speed);
        }

      if (speed.tx.lanes && speed.tx.speed)
        {
          g_print ("   %s %s tx speed:   %u Gb/s = %u lanes * %u Gb/s\n",
                   bolt_glyph (TREE_VERTICAL),
                   tree_branch,
                   speed.tx.lanes * speed.tx.speed,
                   speed.tx.lanes, speed.tx.speed);
        }

      flags = bolt_flags_to_string (BOLT_TYPE_AUTH_FLAGS, aflags, NULL);
      g_print ("   %s %s authflags:  %s\n",
               bolt_glyph (TREE_VERTICAL),
               tree_right,
               flags);
    }

  if (format_timestamp (dev, buf, sizeof (buf), "authtime"))
    g_print ("   %s authorized:    %s\n", tree_branch, buf);

  if (format_timestamp (dev, buf, sizeof (buf), "conntime"))
    g_print ("   %s connected:     %s\n", tree_branch, buf);

  format_timestamp (dev, buf, sizeof (buf), "storetime");
  g_print ("   %s stored:        %s\n", tree_right, buf);

  if (stored)
    {
      const char *pstr = bolt_policy_to_string (policy);
      const char *kstr;

      if (keystate == BOLT_KEY_MISSING)
        kstr = "no";
      else if (keystate == BOLT_KEY_HAVE)
        kstr = "yes";
      else if (keystate == BOLT_KEY_NEW)
        kstr = "yes (new)";
      else
        kstr = "unknown";

      g_print ("   %s %s policy:     %s\n", tree_space, tree_branch, pstr);
      g_print ("   %s %s key:        %s\n", tree_space, tree_right, kstr);

    }

  g_print ("\n");
}


/* ****  */

static SubCommand subcommands[] = {
  {"authorize",    authorize,     "Authorize a device"},
  {"config",       config,        "Get or set global, device or domain properties"},
  {"domains",      list_domains,  "List the active thunderbolt domains"},
  {"enroll",       enroll,        "Authorize and store a device in the database"},
  {"forget",       forget,        "Remove a stored device from the database"},
  {"info",         info,          "Show information about a device"},
  {"list",         list_devices,  "List connected and stored devices"},
  {"monitor",      monitor,       "Listen and print changes"},
  {"power",        power,         "Force power configuration of the controller"},
  {NULL,           NULL,          NULL},
};

char *
subcommands_make_summary (const SubCommand *cmds)
{
  GString *s = g_string_new ("Commands:");
  unsigned int spacing = 0;

  for (const SubCommand *c = cmds; c && c->name; c++)
    spacing = MAX (spacing, strlen (c->name));

  spacing = MAX (15, spacing) + 2;

  for (const SubCommand *c = cmds; c && c->name; c++)
    {
      int space = spacing - strlen (c->name);
      g_string_append_printf (s, "\n  %s", c->name);
      g_string_append_printf (s, "%*s%s", space, "", c->desc);
    }

  return g_string_free (s, FALSE);
}

const SubCommand *
subcommands_find (const SubCommand *cmds,
                  const char       *cmdname,
                  GError          **error)
{
  const SubCommand *cmd = NULL;

  for (const SubCommand *c = cmds; !cmd && c && c->name; c++)
    if (g_str_equal (cmdname, c->name))
      cmd = c;

  if (cmd == NULL)
    {
      g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
                   "Invalid command: %s", cmdname);
      return NULL;
    }

  return cmd;
}

int
subcommand_run (const SubCommand *cmd,
                BoltClient       *client,
                int               argc,
                char            **argv)
{
  g_autofree char *cmdline = NULL;
  g_autofree char **args = NULL;
  int count;

  cmdline = g_strconcat (g_get_prgname (),
                         " ",
                         cmd->name,
                         NULL);

  g_set_prgname (cmdline);

  count = MAX (argc - 1, 1);
  args = g_new0 (char *, count + 1);

  args[0] = cmdline;
  for (int i = 1; i < count; i++)
    args[i] = argv[i + 1];

  return cmd->fn (client, count, args);
}

int
main (int argc, char **argv)
{
  g_autoptr(GOptionContext) optctx = NULL;
  g_autoptr(BoltClient) client = NULL;
  g_autoptr(GError) error = NULL;
  g_autofree char *summary = NULL;
  const SubCommand *cmd = NULL;
  const char *cmdname = NULL;
  const char *uuid_fmtstr = "full";
  gboolean version = FALSE;
  int fmt = -1;
  GOptionEntry options[] = {
    { "version", 0, 0, G_OPTION_ARG_NONE, &version, "Print version information and exit", NULL },
    { "uuids", 'U', 0, G_OPTION_ARG_STRING, &uuid_fmtstr, "How to format uuids [*full, short, alias]", NULL },
    { NULL }
  };

  setlocale (LC_ALL, "");

  optctx = g_option_context_new ("[COMMAND]");
  g_option_context_add_main_entries (optctx, options, NULL);

  summary = subcommands_make_summary (subcommands);
  g_option_context_set_summary (optctx, summary);
  g_option_context_set_strict_posix (optctx, TRUE);

  if (!g_option_context_parse (optctx, &argc, &argv, &error))
    return usage_error (error);

  if (version)
    {
      g_print ("%s %s\n", PACKAGE_NAME, PACKAGE_VERSION);
      exit (EXIT_SUCCESS);
    }

  if (argc < 2)
    cmdname = "list";
  else
    cmdname = argv[1];

  fmt = format_uid_init (uuid_fmtstr, &error);
  if (fmt == -1)
    return usage_error (error);

  client = bolt_client_new (&error);

  if (!client)
    return report_error ("could not create client", error);

  cmd = subcommands_find (subcommands, cmdname, &error);

  if (cmd == NULL)
    return usage_error (error);

  return subcommand_run (cmd, client, argc, argv);
}
